Address :trink's review feedback

mreid-moz opened the pull request:
This addresses feedback from @trink on Pull Request 7.

mreid-moz committed:
Address :trink's review feedback

mreid-moz committed:
Add a lpeg "split path" function

trink commented on heka/sandbox/encoders/http_edge_encoder.lua:
```diff
@@ -4,34 +4,19 @@
-
-function split_path(str)
-    return split(str,'[\\/]+')
+    sep = lpeg.P(sep_char)
```
You won't want to build the grammar for every request

trink commented on heka/sandbox/encoders/http_edge_encoder.lua:
```diff
@@ -68,24 +53,25 @@ local ns_config = {
+    --     ^/submit/namespace/id[/extra/path/components]$
     local path = read_message("Fields[Path]")
-    local components = split_path(path)
+    local components = split_path(path, "/")
```
split_path is now nil

mreid-moz committed:
Pre-generate the lpeg grammar, and always use "/".

trink commented on heka/sandbox/encoders/http_edge_encoder.lua:
```diff
@@ -68,24 +53,25 @@ local ns_config = {
 
     -- Skip this message: Not enough path components.
-    if components == nil or table.getn(components) < 3 then
+    if not components or #components < 3 then
```
components can still never be nil from the current split function (i.e., the test is uncessary)

mreid-moz replied:
Ah, wasn't sure if `lpeg.match(...)` could return nil.

mreid-moz committed:
Don't bother checking components for nil.

mreid-moz committed:
Pull namespace configuration out of the encoder.

Configuration is now expected as a JSON string in the heka config.

mreid-moz committed:
Add a comment + examples for the path PEG.

mreid-moz committed:
Merge branch 'trink_feedback_pr7' into external_ns_config

trink commented on heka/sandbox/encoders/http_edge_encoder.lua:
```diff
@@ -68,24 +61,25 @@ local ns_config = {
 }
 
 function process_message()
-    -- Carry forward payload some incoming fields
```
In general we initialize msg with all the fields we use:
1) to avoid additional allocation/table resizing on the first invocation of process_message
2) to document the message structure (this applies to Fields too although you have a bunch of dynamic ones here)

mreid-moz replied:
We could stop adding dynamic fields at this point - we could simply add a "PathComponents" array regardless of namespace config, and extract those dynamic fields downstream. I don't know if it's desirable to reference (or split) the namespace configuration in multiple places though.

trink commented on heka/sandbox/encoders/http_edge_encoder.lua:
```diff
@@ -99,32 +93,30 @@ function process_message()
+    local num_components = #components
     if num_components > 0 then
         local dims = cfg["dimensions"]
         if dims ~= nil and table.getn(dims) >= num_components then
```
use #dims instead of getn

trink commented on heka/sandbox/encoders/http_edge_encoder.lua:
```diff
@@ -137,11 +129,10 @@ function process_message()
         end
```
Since you are reusing Fields you can leak variables into the next message

mreid-moz committed:
Initialize all the main msg fields we plan to use.

mreid-moz committed:
Add an overly defensive nil check.

If we change the `split_path` function to possibly return nil, we
don't want to be surprised by failures later on.

mreid-moz committed:
Remove extra line

mreid-moz committed:
Stop using table.getn(foo) and start using #foo

trink commented on heka/sandbox/encoders/http_edge_encoder.lua:
```diff
@@ -40,52 +34,56 @@ local UNK_GEO = "??"
     Payload     = nil,
+    EnvVersion  = nil,
+    Hostname    = nil,
     Fields      = {}
```
Just set it to nil since it is a place holder

mreid-moz committed:
Initialize Fields to nil
